demos/TicTacToe/Classes/MainMenuScene.cpp

```diff
@@ -22,33 +76,345 @@ bool MainMenuScene::init() {
+      text_field->attachWithIME();
+    } else {
+      auto text_field = dynamic_cast<TextFieldTTF*>(event->getCurrentTarget());
+      text_field->setCursorEnabled(false);
+      text_field->detachWithIME();
+    }
+
+    return true;
+  };
+
+  auth_background->addChild(password_text_field, 1);
+  auth_background->addChild(password_text_field_border, 1);
+
+  // Attaching the touch listener to the text field.
+  Director::getInstance()
+      ->getEventDispatcher()
+      ->addEventListenerWithSceneGraphPriority(
+          password_text_field_touch_listener, password_text_field);
+  // Creating the login button and giving it a position, anchor point and
+  // touch_listener.
+  auto login_button = Sprite::create(kLoginButtonImage);
+  login_button->setPosition(90, 120);
+  login_button->setAnchorPoint(Vec2(0, 0));
+  login_button->setContentSize(Size(200, 75));
   // Create a button listener to handle the touch event.
-  auto create_button_touch_listener = EventListenerTouchOneByOne::create();
-  // Setting the onTouchBegan event up to a lambda tha will replace the MainMenu
-  // scene with a TicTacToe scene.
-  create_button_touch_listener->onTouchBegan = [](Touch* touch,
-                                                  Event* event) -> bool {
+  auto login_button_touch_listener = EventListenerTouchOneByOne::create();
+  // Setting the onTouchBegan event up to a lambda tha will replace the
+  // MainMenu scene with a TicTacToe scene and pass in login_text_field
+  // string.
+  login_button_touch_listener->onTouchBegan = [this](Touch* touch,
+                                                     Event* event) -> bool {
+    if (previous_state != current_state || current_state != kAuthState) {
+      return true;
+    }
     auto bounds = event->getCurrentTarget()->getBoundingBox();
     auto point = touch->getLocation();
-    // Replaces the scene with a new TicTacToe scene if the touched point is
-    // within the bounds of the button.
     if (bounds.containsPoint(point)) {
-      Director::getInstance()->replaceScene(
-          TicTacToe::createScene(std::string()));
+      if (!std::regex_match(email_text_field->getString(), email_pattern)) {
+        invalid_login_label->setString("invalid email address");
+      } else if (password_text_field->getString().length() < 8) {
+        invalid_login_label->setString(
+            "password must be at least 8 characters long");
+      } else {
+        user_result = auth->SignInWithEmailAndPassword(
+            email_text_field->getString().c_str(),
+            password_text_field->getString().c_str());
+        current_state = kWaitingLoginState;
+      }
     }
+    return true;
+  };
+  // Attaching the touch listener to the login button.
+  Director::getInstance()
+      ->getEventDispatcher()
+      ->addEventListenerWithSceneGraphPriority(login_button_touch_listener,
+                                               login_button);
+  auth_background->addChild(login_button, 1);
```

Comment:
recommendation for readability, the 1 here seems like a magic number.  Instead put a comment before it denoting what the parameter name is.  So say it's like, a layer index:

auth_background->addChild(login_button, /*layer_index=*/1);  

you don't need to do this but it makes nice to read, which is great for a tutorial source file.